spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/log/Slf4jScopeDecorator.java

```diff
@@ -43,196 +41,48 @@
-
-	private final SleuthProperties sleuthProperties;
-
-	private final SleuthSlf4jProperties sleuthSlf4jProperties;
+	private final ScopeDecorator delegate;
 
 	Slf4jScopeDecorator(SleuthProperties sleuthProperties,
 			SleuthSlf4jProperties sleuthSlf4jProperties) {
-		this.sleuthProperties = sleuthProperties;
-		this.sleuthSlf4jProperties = sleuthSlf4jProperties;
-	}
-
-	static void replace(String key, @Nullable String value) {
-		if (value != null) {
-			MDC.put(key, value);
-		}
-		else {
-			MDC.remove(key);
-		}
-	}
-
-	@Override
-	public CurrentTraceContext.Scope decorateScope(TraceContext currentSpan,
-			CurrentTraceContext.Scope scope) {
-		final String previousTraceId = MDC.get("traceId");
-		final String previousParentId = MDC.get("parentId");
-		final String previousSpanId = MDC.get("spanId");
-		final String spanExportable = MDC.get("spanExportable");
-		final String legacyPreviousTraceId = MDC.get(LEGACY_TRACE_ID_NAME);
-		final String legacyPreviousParentId = MDC.get(LEGACY_PARENT_ID_NAME);
-		final String legacyPreviousSpanId = MDC.get(LEGACY_SPAN_ID_NAME);
-		final String legacySpanExportable = MDC.get(LEGACY_EXPORTABLE_NAME);
-		final List<AbstractMap.SimpleEntry<String, String>> previousMdc = previousMdc();
 
-		if (currentSpan != null) {
-			String traceIdString = currentSpan.traceIdString();
-			MDC.put("traceId", traceIdString);
-			MDC.put(LEGACY_TRACE_ID_NAME, traceIdString);
-			String parentId = currentSpan.parentId() != null
-					? HexCodec.toLowerHex(currentSpan.parentId()) : null;
-			replace("parentId", parentId);
-			replace(LEGACY_PARENT_ID_NAME, parentId);
-			String spanId = HexCodec.toLowerHex(currentSpan.spanId());
-			MDC.put("spanId", spanId);
-			MDC.put(LEGACY_SPAN_ID_NAME, spanId);
-			String sampled = String.valueOf(currentSpan.sampled());
-			MDC.put("spanExportable", sampled);
-			MDC.put(LEGACY_EXPORTABLE_NAME, sampled);
-			log("Starting scope for span: {}", currentSpan);
-			if (currentSpan.parentId() != null) {
-				if (log.isTraceEnabled()) {
-					log.trace("With parent: {}", currentSpan.parentId());
-				}
+		CorrelationScopeDecorator.Builder builder = MDCScopeDecorator.newBuilder().clear()
+				.addField(BaggageFields.TRACE_ID).addField(BaggageFields.PARENT_ID)
+				.addField(BaggageFields.SPAN_ID)
+				.addField(BaggageFields.SAMPLED, "spanExportable");
+
+		Set<String> whitelist = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+		whitelist.addAll(sleuthSlf4jProperties.getWhitelistedMdcKeys());
+
+		Set<String> retained = new LinkedHashSet<>();
+		retained.addAll(sleuthProperties.getBaggageKeys());
+		retained.addAll(sleuthProperties.getLocalKeys());
+		retained.addAll(sleuthProperties.getPropagationKeys());
+
+		for (String name : retained) {
+			if (whitelist.contains(name)) {
+				// Until we move off ExtraFieldPropagation onto BaggagePropagation,
+				// manually create the fields...
+				builder.addField(BaggageField.create(name));
+				builder.addDirtyName(name);
```

Comment:
there are some issue history in sleuth about people wanting it to intentionally clobber out-of-band changes. Later configuration can be added to make this far cheaper from an overhead perspective for those not fighting with `MDC.put`